internal/c/parts/video/font/font.cpp

```diff
@@ -1,339 +1,1417 @@
+                FONT_DEBUG_CHECK(bitmap && dst);
+
+                auto dstR = dstL + size.x; // right of dst + 1 where we will end
+                auto dstB = dstT + size.y; // bottom of dst + 1 where we will end
+                auto alphaSrc = bitmap;
+                for (FT_Pos dy = dstT; dy < dstB; dy++) {
+                    for (FT_Pos dx = dstL; dx < dstR; dx++) {
+                        if (dx >= 0 && dx < dstW && dy >= 0 && dy < dstH) { // if we are not clipped
+                            auto dstP = (dst + dstW * dy + dx);             // dst pointer
+                            if (*alphaSrc > *dstP)                          // blend both alpha and save to dst pointer
+                                *dstP = *alphaSrc;
+                        }
+                        ++alphaSrc;
+                    }
+                }
+            }
+        };
+
+        std::unordered_map<uint32_t, Glyph *> glyphs; // holds pointers to cached glyph data for codepoints
+
+        // Delete copy and move constructors and assignments
+        Font(const Font &) = delete;
+        Font &operator=(const Font &) = delete;
+        Font(Font &&) = delete;
+        Font &operator=(Font &&) = delete;
+
+        /// @brief Initializes all members
+        Font() {
+            isUsed = false;
+            fontData = nullptr;
+            face = nullptr;
+            monospaceWidth = defaultHeight = baseline = 0;
         }
-        if (ox < 0) { // compensate for loss of width from left shift
-            w2 = w2 + (-ox);
+
+        /// @brief Frees any cached glyph
+        ~Font() {
+            // Free the FreeType face object
+            if (FT_Done_Face(face)) {
+                FONT_DEBUG_PRINT("Failed to free FreeType face object (%p)", face);
+            } else {
+                FONT_DEBUG_PRINT("FreeType face object freed");
+            }
+
+            // Free the buffered font data
+            free(fontData);
+            FONT_DEBUG_PRINT("Raw font data buffer freed");
+
+            FONT_DEBUG_PRINT("Freeing cached glyphs");
+            // Free any allocated glyph manager
+            // This should also call the glyphs destructor freeing the bitmap data
+            for (auto &it : glyphs)
+                delete it.second;
         }
 
-        // Monospace resize as necessary
-        if (fonts[i].monospace) {
-            if (w2 != fonts[i].monospace_width) {
-                w2 = fonts[i].monospace_width;
-                ox = 0;                // no repositioning possible
-                ox2 = w2 / 2 - w1 / 2; // align to centre
+        /// @brief Creates a glyph belonging to a codepoint, caches its bitmap + info and adds it to the hash map
+        /// @param codepoint A valid UTF-32 codepoint
+        /// @return True if successful or if the glyph is already in the map
+        bool CacheGlyph(FT_ULong codepoint) {
+            if (glyphs.count(codepoint) == 0) {
```

Comment:
God forbid they name it `.exists()` so I don't have to look up what `.count()` is supposed to do here 🤣